pg_search/src/postgres/storage/fsm.rs

```diff
@@ -34,413 +35,1263 @@ enum FSMBlockKind {
 
+            if when_recyclable.value > max_slot.key {
+                // this is safe as the tree still maintains its balance.  we also have an exclusive lock
+                // on the tree at this stage which means no concurrent backends can be changing the tree
+                max_slot.key = when_recyclable.value;
+            }
+
+            max_slot.tag
+        }
+
+        fn extend_blocklist(
+            bman: &mut BufferManager,
+            start_block: BufferMut,
+            mut extend_with: Peekable<impl Iterator<Item = pg_sys::BlockNumber>>,
+        ) -> BufferMut {
+            let mut buffer = start_block;
+            loop {
                 let mut page = buffer.page_mut();
-                let contents = page.contents_mut::<FSMBlock>();
-                let mut cnt = 0;
-                contents
-                    .entries
-                    .iter_mut()
-                    .filter(|FSMEntry(blockno, _)| *blockno == pg_sys::InvalidBlockNumber)
-                    .zip(&mut extend_with)
-                    .for_each(|(entry, blockno)| {
-                        *entry = FSMEntry(blockno, when_recyclable);
-                        cnt += 1;
-                    });
-
-                if cnt == 0 {
-                    // we didn't make any modifications so the page is not dirty -- no need to WAL log it
-                    buffer.set_dirty(false);
-                } else {
-                    // we added at least one block to this page so it's no longer empty
-                    contents.header.empty = false;
+                let contents = page.contents_mut::<AvlLeaf>();
+                let next_blockno = page.next_blockno();
+
+                if contents.len as usize == contents.entries.len()
+                    && next_blockno != pg_sys::InvalidBlockNumber
+                {
+                    // this block is full.  Chances are the next block is also full.
+                    // so we're going to populate a brand-new list with the rest of the iterator
+                    // and then link it in between this block and the next block.  This avoids
+                    // the possible overhead of scanning the rest of the blocklist just to discover
+                    // all the following blocks are full
+                    let new_block = AvlLeaf::init_new_page(bman);
+                    let new_blockno = new_block.number();
+
+                    let mut last_block = Self::extend_blocklist(bman, new_block, extend_with);
+
+                    // link the last block we just created to the next block
+                    let mut end_page = last_block.page_mut();
+                    end_page.special_mut::<BM25PageSpecialData>().next_blockno = next_blockno;
+
+                    // finally link this block to the new block
+                    page.special_mut::<BM25PageSpecialData>().next_blockno = new_blockno;
+
+                    return last_block;
```

Comment:
This seems to bias towards the performance of `extend`, and away from the performance of `drain`, because it will create "fluffier"/less-dense freelists, which will take longer to traverse at `drain` time.